### PR TITLE
feat(rpc): Move away from `serde` in RPC DTO's (in favor of our own `[De]SerializeForVersion`)

### PR DESCRIPTION
**Pull request by t00ts**
This PR is part of https://github.com/eqlabs/pathfinder/issues/2220

### Summary

This PR removes mostly all `serde::Serialize` and `serde::Deserialize` present in our DTO structs, replacing them with ad-hoc implementations of `SerializeForVersion` and `DeserializeForVersion`.

**Note:** For those types that don't necessarily need `Deserialize` capabilities other than in `cfg(test)`, I've tried to keep the `serde` attributes for completeness, although most tests pass without them.

### CODE REVIEW

File: crates/rpc/src/lib.rs [hunk 1]
```diff
@@ -840,11 +859,22 @@ mod tests {
         ];
 
         for (line, input, expected) in examples {
-            let parsed = serde_json::from_str::<Syncing>(input).unwrap();
-            let output = serde_json::to_string(&parsed).unwrap();
+            println!("input: {input}");
+
```
Comment (kkovaacs) [inline crates/rpc/src/lib.rs:863]: I guess this is a leftover debug print?

File: crates/rpc/src/method/call.rs [hunk 2]
```diff
@@ -67,22 +67,32 @@ impl From<CallError> for ApplicationError {
...
+            Ok(Self {
+                contract_address: value.deserialize_serde("contract_address")?,
+                entry_point_selector: value.deserialize_serde("entry_point_selector")?,
+                calldata: value
+                    .deserialize_array("calldata", crate::dto::Value::deserialize_serde)?,
+            })
+        })
+    }
+}
+
```
Comment (kkovaacs) [inline crates/rpc/src/method/call.rs:95]: Does this implementation make most of the `DeserializeForVersion` implementation for `Input` redundant?

File: crates/rpc/src/types/syncing.rs [hunk 3]
```diff
@@ -23,15 +22,36 @@ impl std::fmt::Display for Syncing {
     }
 }
 
+impl crate::dto::serialize::SerializeForVersion for Syncing {
+    fn serialize(
+        &self,
+        serializer: crate::dto::serialize::Serializer,
+    ) -> Result<crate::dto::serialize::Ok, crate::dto::serialize::Error> {
+        match self {
+            Syncing::False(b) => serializer.serialize_bool(*b),
```
Comment (kkovaacs) [inline crates/rpc/src/types/syncing.rs:31]: According to the spec the only possible value here is false, so we could just as well remove the `bool` from our `Syncing` enum now that we can properly serialize that variant into a `false` in JSON.

File: crates/rpc/src/lib.rs (hunk 1 above)
Comment (t00ts) [reply crates/rpc/src/lib.rs:863]: Oops!

File: crates/rpc/src/method/call.rs (hunk 2 above)
Comment (t00ts) [reply crates/rpc/src/method/call.rs:95]: Great catch!

File: crates/rpc/src/types/syncing.rs (hunk 3 above)
Comment (t00ts) [reply crates/rpc/src/types/syncing.rs]: Alright!

**Pull request merged**